### ISSUE TITLE
[GR-44604] Implement Component Model Proposal for WebAssembly

### ISSUE THREAD

**Issue by fniephaus** [oracle/graal#9762]
### TL;DR

We plan to implement [the WebAssembly Component Model proposal](https://github.com/WebAssembly/component-model) in GraalWasm.

### Details

When embedding a WebAssembly module, you get to access its exports, mainly the functions that you will be calling. However, these functions only operate on very simple primitive types (integers and floating-point numbers). Richer types like strings or lists have to be represented in terms of integers and floating-point numbers, for example by the use of pointers and offsets. Those representations are oftentimes not interchangeable across languages.

The Component Model proposal aims to fix this by introducing a new abstraction for composing WebAssembly code - components. Like WebAssembly modules, components define their interface as a set of imports and exports, but unlike WebAssembly modules, only functions can be imported and exported. However, these functions have high-level type definitions and can work on rich data types.
Internally, the component is responsible for mapping these high-level data types onto the WebAssembly primitives of the underlying module. By implementing this proposal, we will improve the interoperability of GraalWasm with Java and other Graal Languages.

More details at https://component-model.bytecodealliance.org/

**Comment by aochagavia** [comment]
Hello there! I stumbled upon this issue as I'm developing a plugin system based on Wasm components. WASI 0.2 has been out there for two years now and I have the impression that the component model is here to stay (WASI 0.3 is [around the corner](https://github.com/bytecodealliance/wasi.dev/blob/cfc6d0347063f397aff6bf264a5e9666993855a6/docs/roadmap.md) and is building upon all the good ideas in WASI 0.2, including the component model). Do you think it would make sense to prioritize this short term? I'd love to use Graal Wasm to run Wasm components from a JVM host.

**Comment by woess** [comment]
@aochagavia WASI 0.3 and the component model have been around the corner for [a](https://github.com/bytecodealliance/wasi.dev/commit/35ed368586416073024b1ce4e5e3a2102cb31962) [while](https://github.com/bytecodealliance/wasi.dev/commit/66b57ebc6d3447e84f9a26a14ff2b3a97e578304) now... 😅
So no, short term, we're going to prioritize WebAssembly 3.0 completeness and other features. Once that's done and WASI 0.3 is released, we'll look into implementing it.
